README.md
 ```
    _____             _               __  __                        _        
   |  __ \           | |             |  \/  |                      | |       
   | |  | | ___   ___| | _____ _ __  | \  / | __ _  __ _  ___ _ __ | |_ ___  
   | |  | |/ _ \ / __| |/ / _ \ '__| | |\/| |/ _` |/ _` |/ _ \ '_ \| __/ _ \ 
   | |__| | (_) | (__|   <  __/ |    | |  | | (_| | (_| |  __/ | | | || (_) |
   |_____/ \___/ \___|_|\_\___|_|    |_|  |_|\__,_|\__, |\___|_| |_|\__\___/ 
                                                    __/ |                    
                                                   |___/                   
```
# Docker development environment for Magento 1 and Magento 2

## Services
* `elasticsearch`: [docker.elastic.co/elasticsearch/elasticsearch:7.6.2](https://github.com/danhort/docker-magento2/blob/main/elasticsearch/Dockerfile)
* `maildev`: [djfarrelly/maildev:latest](https://hub.docker.com/r/djfarrelly/maildev/)
* `mysql`: [mysql:5.7](https://store.docker.com/images/mysql)
* `nginx`: [nginx:1.15-alpine](https://github.com/danhort/docker-magento2/blob/main/nginx/Dockerfile)
* `php` :
    - [php:5.6-fpm-alpine](https://github.com/danhort/docker-magento2/blob/main/php/5.6/Dockerfile)
    - [php:7.0-fpm-alpine](https://github.com/danhort/docker-magento2/blob/main/php/7.0/Dockerfile)
    - [php:7.1-fpm-jessie](https://github.com/danhort/docker-magento2/blob/main/php/7.1/Dockerfile)
    - [php:7.2-fpm-alpine](https://github.com/danhort/docker-magento2/blob/main/php/7.2/Dockerfile)
    - [php:7.3-fpm-alpine](https://github.com/danhort/docker-magento2/blob/main/php/7.3/Dockerfile)
    - [adhocore/phpfpm:7.4](https://github.com/danhort/docker-magento2/blob/main/php/7.3=4/Dockerfile)
* `redis`: [redis:4-alpine](https://store.docker.com/images/redis)
* `varnish`: [varnish](https://hub.docker.com/_/varnish
* `phpmyadmin`: [phpmyadmin/phpmyadmin](https://github.com/danhort/docker-magento2/blob/main/phpmyadmin/Dockerfile)

## Setup
In your project root dir.
```
wget https://raw.githubusercontent.com/danhort/docker-magento/main/bin/init.sh
chmod +x init.sh
./init.sh
```
This creates a Makefile in the project root, a docker.env file and a custom-nginx.conf file in the ops directory.  
Update the docker.env variables COMPOSE_PROJECT_NAME, MAGENTO_ROOT, PHP_VERSION, MAGENTO_VERSION.   
Update the ops/custom-nginx.conf file with location directives you need. (e.g. blog/)   
Create the source folder (MAGENTO_ROOT) in the project root where your code will live.

## Install docker-magento tool (optional)
Install docker-magento binary to access the init script from everywhere.   
This will give you a bash tool to setup this environment using the command `docker-magento`
```
make install
```

## Usage
```
make
```

## Links
magento: https://magento.localhost  
magento admin: https://magento.localhost/admin   
maildev: http://localhost:1080  
phpMyAdmin: http://localhost:8080

## Database credentials
host: mysql   
name: magento   
user: magento   
password: magento   

There is a Magento 2 env.php sample file in ops/samples

## Xdebug
To enable Xdebug you must change the `DOCKER_PHP_IMAGE` variable in ops/docker.env from `magento_php` to `magento_php_xdebug`.   
The port used for Xdebug is 9001.   
The following is a configuration sample for vscode: ops/samples/launch.json.sample

## Comments
*This environment setup has only been tested on linux.*


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.8.0] - 2021-05-27
### Added
- Ioncube to PHP 4
### Fixed
- Added sed command to replace utf8mb4_0900_ai_ci to prevent errors in db import

## [0.7.1] - 2021-04-19
### Added
- Fixed symlinking external vendor module

## [0.7.0] - 2021-04-19
### Added
- Added external composer extensions volume + symlink command

## [0.6.0] - 2021-04-07
### Added
- Git to PHP 7.4
### Changed
- Update composer to v2 in PHP 7.4

## [0.5.8] - 2021-03-12
### Changed
- Replace zcat with gunzip for macOS compatibility
- Update db-import command to recreate database table
### Fixed
- Change user in db-dump command

## [0.5.7] - 2021-01-14
### Fixed
- Fix composer removal error

## [0.5.6] - 2021-01-13
### Fixed
- Makefile missing space

## [0.5.5] - 2021-01-10
### Changed
- Changed master references to main in readme

## [0.5.4] - 2021-01-10
### Fixed
- Downgrade composer to version 1 (Magento does not support composer 2)

## [0.5.3] - 2020-12-18
### Fixed
- Makefile db import and export bug

## [0.5.2] - 2020-11-10
### Added
- patch library ro PHP7.4 

## [0.5.1] - 2020-11-10
### Fixed
- Makefile conflict error

## [0.5.0] - 2020-10-30
### Added
- Varnish
- Ioncube to PHP 7.3 container
- Add base_static_url and base_media_url and elastic host to env.php sample
- Add fastcgi_buffer_size to nginx
- Added PHP 7.4
### Changed
- Updated elasticsearch to 7.6
### Removed
- Remove unused bin script

## [0.4.0] - 2020-05-19
### Added
- npm to PHP 7.0, 7.1, 7.2 and 7.3
- npm and grunt commands to makefile

## [0.3.0] - 2020-05-18
### Fixed
- Swapped clear-assets src path with env path
### Changed
- Removed www from env sample default urls
### Added
- config file update method to update bin script

## [0.2.0] - 2020-05-06
### Added
- PHP7.3

## [0.1.0] - 2020-04-29
### Changed
- The update command now checks for the latest tag
- The magento-docker bash tool now initiates the latest tag

## [0.0.1] - 2020-04-29
### Added
- This CHANGELOG file to hopefully serve as an evolving example of a
  standardized open source project CHANGELOG.

[Unreleased]: https://github.com/danhort/docker-magento/compare/0.8.0...HEAD
[0.8.0]: https://github.com/danhort/docker-magento/compare/0.7.1...0.8.0
[0.7.1]: https://github.com/danhort/docker-magento/compare/0.7.0...0.7.1
[0.7.0]: https://github.com/danhort/docker-magento/compare/0.6.0...0.7.0
[0.6.0]: https://github.com/danhort/docker-magento/compare/0.5.8...0.6.0
[0.5.8]: https://github.com/danhort/docker-magento/compare/0.5.7...0.5.8
[0.5.7]: https://github.com/danhort/docker-magento/compare/0.5.6...0.5.7
[0.5.6]: https://github.com/danhort/docker-magento/compare/0.5.5...0.5.6
[0.5.5]: https://github.com/danhort/docker-magento/compare/0.5.4...0.5.5
[0.5.4]: https://github.com/danhort/docker-magento/compare/0.5.3...0.5.4
[0.5.3]: https://github.com/danhort/docker-magento/compare/0.5.2...0.5.3
[0.5.2]: https://github.com/danhort/docker-magento/compare/0.5.1...0.5.2
[0.5.1]: https://github.com/danhort/docker-magento/compare/0.5.0...0.5.1
[0.5.0]: https://github.com/danhort/docker-magento/compare/0.4.0...0.5.0
[0.4.0]: https://github.com/danhort/docker-magento/compare/0.3.0...0.4.0
[0.3.0]: https://github.com/danhort/docker-magento/compare/0.2.0...0.3.0
[0.2.0]: https://github.com/danhort/docker-magento/compare/0.1.0...0.2.0
[0.1.0]: https://github.com/danhort/docker-magento/compare/0.0.1...0.1.0
[0.0.1]: https://github.com/danhort/docker-magento/releases/tag/0.0.1


php/7.2/users/env.sample.php
<?php
return [
    'backend' => [
        'frontName' => 'admin'
    ],
    'crypt' => [
        'key' => 'e8a4cdb6309df49f2969b7342f8a6458'
    ],
    'db' => [
        'table_prefix' => '',
        'connection' => [
            'default' => [
                'host' => 'mysql',
                'dbname' => 'magento',
                'username' => 'magento',
                'password' => 'magento',
                'active' => '1'
            ]
        ]
    ],
    'resource' => [
        'default_setup' => [
            'connection' => 'default'
        ]
    ],
    'x-frame-options' => 'SAMEORIGIN',
    'MAGE_MODE' => 'developer',
    'session' => [
        'save' => 'redis',
        'redis' => [
            'host' => 'redis',
            'port' => '6379',
            'password' => '',
            'timeout' => '2.5',
            'persistent_identifier' => '',
            'database' => '2',
            'compression_threshold' => '2048',
            'compression_library' => 'gzip',
            'log_level' => '1',
            'max_concurrency' => '6',
            'break_after_frontend' => '5',
            'break_after_adminhtml' => '30',
            'first_lifetime' => '600',
            'bot_first_lifetime' => '60',
            'bot_lifetime' => '7200',
            'disable_locking' => '0',
            'min_lifetime' => '60',
            'max_lifetime' => '2592000'
        ]
    ],
    'cache_types' => [
        'config' => 1,
        'layout' => 1,
        'block_html' => 1,
        'collections' => 1,
        'reflection' => 1,
        'db_ddl' => 1,
        'compiled_config' => 1,
        'eav' => 1,
        'customer_notification' => 1,
        'config_integration' => 1,
        'config_integration_api' => 1,
        'full_page' => 1,
        'config_webservice' => 1,
        'translate' => 1,
        'vertex' => 1
    ],
    'install' => [
        'date' => 'Thu, 13 Dec 2018 19:27:28 +0000'
    ],
    'cache' => [
        'frontend' => [
            'default' => [
                'backend' => 'Cm_Cache_Backend_Redis',
                'backend_options' => [
                    'server' => 'redis',
                    'port' => '6379',
                    'persistent' => '',
                    'database' => '0',
                    'password' => '',
                    'force_standalone' => '0',
                    'connect_retries' => '1',
                    'read_timeout' => '10',
                    'automatic_cleaning_factor' => '0',
                    'compress_data' => '1',
                    'compress_tags' => '1',
                    'compress_threshold' => '20480',
                    'compression_lib' => 'gzip',
                    'use_lua' => '0',
                ],
            ],
            'page_cache' => [
                'backend' => 'Cm_Cache_Backend_Redis',
                'backend_options' => [
                    'server' => 'redis',
                    'port' => '6379',
                    'persistent' => '',
                    'database' => '1',
                    'password' => '',
                    'force_standalone' => '0',
                    'connect_retries' => '1',
                    'lifetimelimit' => '57600',
                    'compress_data' => '0',
                ],
            ],
        ],
    ],
];
